Prompt to update

damienissa opened the pull request:
Added feature that will show prompt to update message when a new version of Serverpod CLI is available 

#611 

## Message:
```
┌───────────────────────────────────────────────────────────────────────────────────────┐
│ A new version 1.0.0 of Serverpod is available!                                        │
│                                                                                       │
│ To update to the latest version, run "serverpod upgrade".                             │
│ Also, do not forget to update packages in your server, client, and flutter projects.  │
└───────────────────────────────────────────────────────────────────────────────────────┘
```

## Pre-launch Checklist

- [x] I read the [Contribute](https://docs.serverpod.dev/contribute) page and followed the process outlined there for submitting PRs.
- [x] This update contains only one single feature or bug fix and nothing else. (If you are submitting multiple fixes, please make multiple PRs.)
- [x] I read and followed the [Dart Style Guide](https://dart.dev/guides/language/effective-dart/style) and formatted the code with [dart format](https://dart.dev/tools/dart-format).
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`), and made sure that the documentation follows the same style as other Serverpod documentation. I checked spelling and grammar.
- [x] I added new tests to check the change I am making.
- [x] All existing and new tests are passing.
- [ ] Any breaking changes are documented below.

damienissa commented:
I have read the CLA Document and I hereby sign the CLA

yahu1031 commented on tools/serverpod_cli/bin/util/promt_to_update.dart:
```diff
@@ -0,0 +1,62 @@
+''';
+
+/// Check current Serverpod CLI version and promt user to update if needed
+Future<void> promtToUpdateIfNeeded() async {
```
Hope we can give the latest version in the cli for better info.

```console
A new version-1.0.1 of Serverpod is available!
```

something like this. LMK if this is okie, so I can share the rest info(API Endpoints) of pub.dev.

yahu1031 replied:
Or How about writing a command/args like `serverpod upgrade` - to upgrade the serverpod cli version and `serverpod update` (in the project) to update the serverpod deps in the project?

yahu1031 replied:
If possible, try [this](https://sourceb.in/FHYPUpPdcK)

damienissa replied:
It's a good idea. But I'm not sure about updating dependencies. Because a user can call `serverpod upgrade` out of the workspace.

damienissa commented:
@yahu1031  @vlidholt 
Added ability to call `serverpod upgrade` to upgrade CLI to the last available version.
Also updated prompt message:
```
┌───────────────────────────────────────────────────────────────────────────────────────┐
│ A new version 1.0.0 of Serverpod is available!                                        │
│                                                                                       │
│ To update to the latest version, run "serverpod upgrade".                             │
│ Also, do not forget to update packages in your server, client, and flutter projects.  │
└───────────────────────────────────────────────────────────────────────────────────────┘
```

yahu1031 commented on tools/serverpod_cli/bin/util/prompt_to_update.dart:
```diff
@@ -85,6 +63,27 @@ Future<void> updateCLI(
   }
 }
 
+YamlMap? _getcurrentPackage() {
```
Have you tested it? Cause I'm not sure how this will work with `pubspec.yaml` I mean which yaml is it going to pick up?

damienissa replied:
Of course, I tested it. It works well. You can easily test it by running bin/serverpod_cli.dart in debug mode.

damienissa replied:
Also, you might be missing that it has a check `if (contents is! YamlMap) return null;`. That will help to prevent issues with the wrong type.

damienissa commented:
@vlidholt Hello. What about this one?

yahu1031 commented:
Please check the dart analyze issue @damienissa and this is good to go.

damienissa commented:
@yahu1031 Fixed.

vlidholt commented on tools/serverpod_cli/bin/util/prompt_to_update.dart:
```diff
@@ -0,0 +1,131 @@
+  return contents;
+}
+
+String? _getPackageVersion() {
```
The current version is already saved in the `templateVersion` global variable, it would be much better to use that instead.

vlidholt commented on tools/serverpod_cli/bin/util/prompt_to_update.dart:
```diff
@@ -0,0 +1,131 @@
+  });
+}
+
+String _promptMessage(String version) {
```
This feels a bit like a hack. I would either use that other package or write this in a more reusable way.

vlidholt reviewed (changes requested):
Thanks for the PR. :)

Apart from the comments, these things to be fixed, tested, and considered for this:

1. What happens if the user doesn't have access to internet when running the tooling (e.g. works from an airplane).
2. How does this affect performance when running the tooling?

yahu1031 commented:
The answer for Q1 is that this function is only being triggered from the upgrade command, So If the user is offline, We going to get values null at [this line](https://github.com/serverpod/serverpod/blob/25c9bbb1518a8286233f7bce158afd3069ed69e3/tools/serverpod_cli/bin/util/prompt_to_update.dart#L14) and it returns null and the function ends with nothing. I can suggest(@damienissa) have an already up-to-date prompt there.

vlidholt commented:
This is merged in another PR. :)
